Rare segfault when reading the DGA.csv file

Code example:
```
>>> import datatable as dt
>>> for i in range(100):
...   df = dt.fread("../h2o-3/bigdata/laptop/DGA.csv")
...   df.internal.check()
...   del df
... 
Reading file: |█████████████████████       |  88%Segmentation fault: 11
```

Ocasionally however, it would stop with
```
Traceback (most recent call last):
  File "temp/z.py", line 6, in <module>
    df.internal.check()
AssertionError: String offset in row 854015 of Column 0 cannot be less than the previous offset: offset = 0, previous offset = 9282874
```

<details>
<summary>Verbose output from fread:</summary>

```
[1] Prepare for reading
  Input is assumed to be a file name.
  Using default 8 threads
  sep = <auto-detect>
  Decimal separator = '.'
  Quote char = (")
  show_progress = True
  na_strings = ["NA"]
  strip_whitespace = True
  skip_blank_lines = False
  File "../h2o-3/bigdata/laptop/DGA.csv" opened, size: 269695435
  ==== file sample ====
  domain,class,pctCoverage,-,0,1,2,3,4,5,6,7,8,9,_,a,b,c,d,e,f,g,h,i,j,k,l,m,n,o,p,q,r,s,t,u,v,w,x,...
  000directory,0,0.320512820513,0.0,0.25,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.08333333...
  000webhost,0,0.236363636364,0.0,0.3,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.1,0.0,0.0,0.1,0...
  001fans,0,0.25,0.0,0.285714285714,0.142857142857,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.1428571428...
  01-telecharger,0,0.228571428571,0.0714285714286,0.0714285714286,0.0714285714286,0.0,0.0,0.0,0.0,0...
  =====================
  LF character (\n) found in input, \r-only newlines will not be recognized
[2] Detect separator, quoting rule, and ncolumns
  sep=',' with 100 lines of 45 fields using quote rule 0
  Detected 45 columns
  sep = ','
  Quote rule = 0
[3] Detect column types and header
  The input will be read in 1032 chunks of size 261332 each
  Number of sampling jump points = 101 because the first chunk was 9515.4 times smaller than the entire file
  Type codes (jump 000): sbFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFiiF
  `header` determined to be True due to column 2 containing a string on row 1 and type Bool8/numeric in the rest of the sample.
  =====
  Sampled 100 rows (handled \n inside quoted fields) at 101 jump point(s)
  Bytes from first data row to the end of last row: 269695435
  Line length: mean=286.27 sd=67.56 min=181 max=458
  Estimated number of rows: 269695435 / 286.27 = 942102
  Initial alloc = 1490030 rows (942102 + 58%) using bytes/max(mean-2*sd,min) clamped between [1.1*estn, 2.0*estn]
  All rows were sampled since file is small so we know nrows=99 exactly
[4] Assign column names
[5] Apply user overrides on column types
  After 0 type and 0 drop user overrides : sbFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFiiF
  Allocating 45 column slots with 99 rows
[6] Read the data
  The input will be read in 944 chunks of size 285694 each
Reading file: |                                 |   0%  Too few rows allocated, reallocating to 854131 rows
Reading file: |██████████████████████████       |  88%  Too few rows allocated, reallocating to 1161520 rows
[7] Finalize the datatable                                                      
  =============================
  Read 967,798 rows x 45 columns from 257.2MB input in 00:00.489s
   = 0.008s ( 2%) memory-mapping input file
   + 0.002s ( 0%) detecting parse parameters
   + 0.001s ( 0%) detecting column types using 99 sample rows
   + 0.000s ( 0%) allocating [99 x 45] frame (38.37KB) of which 967,798 (977574%) rows used
   + 0.478s (98%) reading data using 8 threads
      = 0.382s (78%) reading into row-major buffers
      + 0.069s (14%) saving into the output frame
      + 0.028s ( 6%) waiting
   + 0.000s ( 0%) creating the final Frame
```
</details>

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -26,6 +26,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 - fix bug where max_nrows parameter was sometimes causing a seg.fault
 - fix fread performance bug caused by memory-mapped file being accidentally
   copied into RAM.
+- fix rare crash in fread when resizing the number of rows.
 
 
 ### [v0.6.0](https://github.com/h2oai/datatable/compare/v0.6.0...v0.5.0) — 2018-06-05

--- a/c/csv/chunks.cc
+++ b/c/csv/chunks.cc
@@ -277,7 +277,7 @@ void ChunkedDataReader::realloc_output_columns(size_t ichunk, size_t new_alloc)
   nrows_allocated = new_alloc;
   g.trace("Too few rows allocated, reallocating to %zu rows", nrows_allocated);
 
-  dt::shared_lock(shmutex, /* exclusive = */ true);
+  dt::shared_lock lock(shmutex, /* exclusive = */ true);
   g.columns.set_nrows(nrows_allocated);
 }
 

--- a/c/csv/reader_fread.cc
+++ b/c/csv/reader_fread.cc
@@ -650,7 +650,7 @@ void FreadReader::detect_column_types()
       trace("Initial alloc = %zd rows (%zd + %d%%) using bytes/max(mean-2*sd,min) clamped between [1.1*estn, 2.0*estn]",
             allocnrow, estnrow, static_cast<int>(100.0*allocnrow/estnrow-100.0));
     }
-    if (tch == eof) {
+    if (nChunks == 1 && tch == eof) {
       if (header == 1) n_sample_lines--;
       estnrow = allocnrow = n_sample_lines;
       trace("All rows were sampled since file is small so we know nrows=%zd exactly", estnrow);

